micro-rdk/src/common/webrtc/udp_mux.rs

```diff
@@ -0,0 +1,578 @@
+        Ok(self.read_header(buf))
+    }
+
+    fn read_header(&self, hdr: [u8; 13]) -> (u16, usize) {
+        let msg_type = hdr[0];
+        if msg_type < 2 {
+            // stun message
+            let len: u16 = u16::from_be_bytes(hdr[2..4].try_into().unwrap());
+            (len, STUN)
+        } else {
+            // assume DTLS record
+            let len: u16 = u16::from_be_bytes(hdr[11..13].try_into().unwrap());
+            (len, DTLS)
+        }
+    }
+    async fn send_to(&self, buf: &[u8], peer: SocketAddr) -> Result<usize> {
+        loop {
+            let socket = self.socket.as_ref().get_ref();
+            match socket.send_to(buf, peer) {
+                Err(e)
+                    if e.kind() == std::io::ErrorKind::WouldBlock
+                        || e.kind() == std::io::ErrorKind::OutOfMemory => {}
+                Ok(s) => return Ok(s),
+                Err(e) => return Err(e),
+            }
+            self.socket.writable().await?;
+        }
+    }
+
+    fn poll_recv_from(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        dir: usize,
+        buf: &mut [u8],
+    ) -> Poll<Result<(usize, SocketAddr)>> {
+        self.register_waker(cx.waker().clone(), dir);
+        loop {
+            let r = match self.peek() {
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => (0, 0),
+                Ok(s) => s,
+                Err(e) => return Poll::Ready(Err(e)),
+            };
+            if r.0 != 0 {
+                if dir == r.1 {
+                    let socket = self.socket.as_ref().get_ref();
+                    self.deregister_waker(dir);
+                    return Poll::Ready(socket.recv_from(buf));
+                }
+
+                match self.yield_or_discard(r.1, r.0) {
+                    Err(e) => return Poll::Ready(Err(e)),
+                    Ok(s) => {
+                        if s {
+                            continue;
+                        }
+                    }
+                }
+            }
+            let _ = ready!(self.socket.poll_readable(cx));
+        }
+    }
+    fn poll_send_to(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+        peer: SocketAddr,
+    ) -> Poll<Result<usize>> {
+        loop {
+            let socket = self.socket.as_ref().get_ref();
+            match socket.send_to(buf, peer) {
+                Err(e)
+                    if e.kind() == std::io::ErrorKind::WouldBlock
+                        || e.kind() == std::io::ErrorKind::OutOfMemory => {}
+
+                Ok(s) => return Poll::Ready(Ok(s)),
+                Err(e) => return Poll::Ready(Err(e)),
+            }
+            let _ = ready!(self.socket.poll_writable(cx));
+        }
+    }
+}
+
+#[derive(Default)]
+struct MuxState {
+    waker: Option<Waker>, // waker is present if a consumer has yield because it's waiting it's turn on the socket
+    interest: bool,       // whether a consumer has a read interest
```

Comment:
the read interest flag is set when the consumer has been taken. The goal behind it is to avoid stalling  the pipeline if for example we get STUN messages expecting  DTLS but there are no STUN reader running

Reply:
I still think something like "consumer_available" with the comment "if false, an existing consumer is expecting to read from this mux" would be clearer.

Reply:
agreed i renamed it to `is_listening`